src/python/pants/core/goals/lint.py

```diff
@@ -160,6 +160,11 @@ def rules():
     is_formatter: ClassVar[bool] = False
     is_fixer: ClassVar[bool] = False
 
+    # Enables lint rules which run the formatter/fixer and check if it made changes.
+    # If it did make changes shows an error. This makes sure the user didn't forget
+    # to run the formatter/fixer. You most probably don't want to disable this flag.
+    enable_lint_rules: ClassVar[bool] = True
```

Comment:
I would expect this on `AbstractFixRequest`, not on `AbstractLintRequest`... (you'd omit the call to `super()._get_rules()` I think.

Also, for the comment, something like:

```
# Enable support for re-using this request's rule in `lint`, where the success/failure of the linter corresponds to 
# whether the rule's output matches the input (i.e. whether the tool made changes or not).
#
# If you set this to `False`, you'll need to provide the following `UnionRule` with a custom class, 
# as well as their corresponding implementation rules: 
#   - `UnionRule(AbstractLintRequest, cls)`
#   - `UnionRule(AbstractLintRequest.Batch, cls)`
#
# !!! Setting this to `False` should be exceedingly rare, as the default implementation handles two important things:
#   - Re-use of the exact same process in `fix` as in `lint`, so runs like `pants fix lint` use 
#     cached/memoized results in `lint`. This pattern is commonly used by developers locally.
#   - Ensuring that `pants lint` is checking that the file(s) are actually fixed. It's easy to forget to provide the 
#     `lint` implementation (which is used usually in CI, as opposed to `fix`), which allows files to be merged 
#     into the default branch un-fixed. (Fun fact, this happened in the Pants codebase before this inheritance existed
#     and was the catalysts for this design).
# The case for disabling this is when the `fix` implementation fixes a strict subset of some `lint` implementation, where
# the check for is-this-fixed in the `lint` implementation isn't possible.
# As an example, let's say tool `cruft` has `cruft lint` which lints for A, B and C. It also has `cruft lint --fix` which fixes A.
# Tthere's no way to not check for `A` in `cruft lint`. Since you're already going to provide a `lint` implementation 
# which corresponds to `cruft lint`, there's no point in running `cruft check --fix` in `lint` as it's already covered by 
# `cruft lint`.
```